Adds MinGW gcc and Windows target for Ztunnel

grnmeira opened the pull request:
This PR adds dependencies that allow us to cross compile ztunnel for Windows (using MinGW).

istio-testing commented:
@grnmeira: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
containers-test_tools | 85ffb93928fb0508c88d1c2585b49d2d902edf8c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_tools/3238/containers-test_tools/1940374408514244608) | true | `/test containers-test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

grnmeira commented:
> what is the image size impact?

The rust target adds 130MB, the MinGW toolchain adds 530MB. We went from a 5.63GB image to a 6.34GB one.

zirain commented:
> > what is the image size impact?
> 
> @howardjohn The rust target adds 130MB, the MinGW toolchain adds 530MB. We went from a 5.63GB image to a 6.34GB one.

Can make it as a serparated one, does everyone need this? it's hard for someone work behind VPN?

grnmeira commented:
An alternative would be to change the permissions of "/home/.rustup" to 755 too. That would allow the tools to be installed during the compilation process, at least for now as the branch is experimental.

grnmeira commented:
@keithmattix @howardjohn @zirain 

My apologies, I just added a third approach here. I've added an option with a --build-arg that can switch between base images, and generate different images. This way we can optionally generate an image that contains the necessary deps for cross compilation.

@keithmattix this simplifies the changes in the istio repo, one can then call "make docker" with an IMG=<hub/image> env variable.

grnmeira commented:
@howardjohn let me know if we can get this merged. Happy to change anything if needed.

dhawton commented:
> @howardjohn let me know if we can get this merged. Happy to change anything if needed.

there is an outstanding question from Keith

dhawton commented:
> Ideally we'd have it in the GCR just like the other images. But I don't think the location of the image should be a blocker for now (unless we need changes in code support that).

If you need windows builds for tests now, then the image should exist.

dhawton commented on docker/build-tools/Dockerfile:
```diff
@@ -36,6 +38,7 @@ ARG TARGETARCH
 # Istio tools SHA that we use for this image
 ARG ISTIO_TOOLS_SHA
 
+
```
Unneeded extra line?

keithmattix commented on docker/build-tools/build-and-push.sh:
```diff
@@ -90,13 +90,25 @@ ${CONTAINER_CLI} ${CONTAINER_BUILDER} --target build_env_proxy \
     "${HUB}/build-tools:${BRANCH}-latest"
     "${HUB}/build-tools-proxy:${VERSION}"
     "${HUB}/build-tools-proxy:${BRANCH}-latest"
+    "${HUB}/build-tools-mingw:${VERSION}"
```
nit: can we just name this windows instead of mingw? It's more searchable

grnmeira replied:
I get what you mean. Though wouldn't it cause confusion, as if the image was running Windows?

istio-testing commented:
@grnmeira: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
containers-test_tools | 8a11eccbe1b97d8272d9d9d5d2d66784a4685150 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_tools/3238/containers-test_tools/1942242895293583360) | true | `/test containers-test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

keithmattix commented:
/test containers-test
